### ISSUE TITLE
Account for paths when installed to the system

### ISSUE THREAD

**Issue by Unit193** [drwetter/testssl.sh#1640]
Howdy,

It's be great if testssl.sh could account for when the binary is installed as /usr/bin/testssl.sh and the configuration lives in /etc/testssl/.  Right now, it would check (among other places) /usr/etc/ which of course wouldn't be right.

In Debian I added an ugly patch[0] so it'd at least pick up the files.

I was also looking into running the offline tests, but in order to do so I'd also have to change the binary and configuration paths.  I was wondering if perhaps the tests could check if $TESTSSL_BINARY/$TESTSSL_ETCDIR were set, and if so use them instead of ~/etc and ./testssl.sh.  This would just be icing on the cake as far as I'm concerned though.

Thanks!

[0]: https://salsa.debian.org/pkg-security-team/testssl.sh/-/blob/debian/master/debian/patches/allow_loading_config.patch

~Unit 193
Unit193 @ freenode
Unit193 @ OFTC

**Comment by drwetter** [comment]
> . I was wondering if perhaps the tests could check if $TESTSSL_BINARY/$TESTSSL_ETCDIR were set, and if so use them instead of ~/etc and ./testssl.sh. This would just be icing on the cake 

I think a separate TESTSSL_ETCDIR would do good. Looking at your patch that would only mean one place which would need to be changed or preset with another default value?

Looking forward and while you were asking and are "here" :-)  : I thought at some point of time splitting the program into main and some other files containing necessary parts shell of the script code.
 
From the perspective of a github user I have a good idea where to put those files. But where would a distribution put those,  ``/usr/lib/testssl/`` ? Wasn't there a standard / formal agreement on that?

cc @chkr-private

**Comment by AJCxZ0** [comment]
There is one: the Linux [Filesystem Hierarchy Standard](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/index.html), however since different distributions have their own local variants, it's usually best to just make it easy for the packagers to put things in the right place just as with e.g. the [FreeBSD port](https://svnweb.freebsd.org/ports/head/security/testssl.sh/).
An excellent cross-Linux-platform option for packaging is _flatpack_. It would be great to get [testssl on Flathub](https://flathub.org/apps/search/testssl).
There is also [snap](https://snapcraft.io/).

**Comment by drwetter** [comment]
Yeah, FHS was the name I was looking for. Thanks, @AJCxZ0 , never heard about flathub. For snap this kind of confused me: https://snapcraft.io/first-snap.

So looking at FHS my suggestion would be 

- introduce ``$TESTSSL_ETC_DIR`` or ``$TESTSSL_ETC``
- define the variable in the beginning with ``TESTSSL_ETC_DIR="${TESTSSL_ETC_DIR:-""}"`` which maybe set differently by distibutors
- then in ``get_install_dir()`` check whether ``TESTSSL_ETC_DIR`` is not empty and ``$TESTSSL_ETC_DIR/cipher-mapping.txt`` can be found.
- replace all other occurrences of  ``TESTSSL_INSTALL_DIR/etc/`` by ``TESTSSL_ETC_DIR``

Would that help, @unit193 ?

Looking forward then I suggest   ``TESTSSL_LIBS`` which will be empty for now. In the future we may put parts of the code there. For github ``TESTSSL_LIBS``  will refer to ``$TESTSSL_INSTALL_DIR/lib``.  Distributors may also here preset  ``TESTSSL_LIBS="${TESTSSL_LIBS:-""}"`` with another value like ``/usr/lib/testssl/``

**Comment by Unit193** [comment]
My apologies for a late response.

With regards to libraries, /usr/lib/testssl/ makes the most sense if you're going to source them, while libexec (/usr/libexec/testssl/) would make more sense if you plan to execute them.

That would help tremendously!  It'd be great if I could do that without patching, but minimal patches are far better than me trying to kludge this in (as I'm doing now), and you want to be able to run testssl.sh directly from the git repo so m4 is out too.

With regards to my idea with `TESTSSL_BINARY` in tests, that's related to the tests expecting `./testssl.sh` to exist, while when we would be running it the package would be installed so `/usr/bin/testssl` would be our location.  That being said, it's just icing on the cake, but I figured I should explain a little.

Thanks for being receptive!

~Unit 193

**Comment by drwetter** [comment]
No problem, my time is pretty limited right now, too.

For the include files: The idea was to just introduce the variable. You just change a line and preset it. That's it.

For ``TESTSSL_BINARY``: off the top of my head don't think that is needed for distributors. We need the installdir pointing to it,for people pulling the source from here.
